### PR TITLE
HIVE-19248: REPL LOAD couldn't copy file from source CM path and also doesn't throw error if file copy fails.

### PR DESCRIPTION
**Pull request by sankarh**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java [hunk 1]
```diff
@@ -82,71 +82,126 @@ public void copyAndVerify(FileSystem destinationFs, Path destRoot,
...
-        pathList = new ArrayList<>();
+        // If copy fails, fall through the retry logic
+        isCopyError = true;
+      }
+      pathList = getFilesToRetry(sourceFs, srcFileList, destinationFs, destination, isCopyError);
+      repeat++;
+    }
+
+    // If files copy fail after several attempts, then throw error
+    if (!pathList.isEmpty() || isCopyError) {
```
Comment (maheshk114) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java]: in what scenario path list will be empty with copy error set to true ?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java [hunk 2]
```diff
@@ -82,71 +82,126 @@ public void copyAndVerify(FileSystem destinationFs, Path destRoot,
...
   }
 
-  private void doCopyRetry(FileSystem sourceFs, List<ReplChangeManager.FileInfo> fileList,
+  private void doCopyRetry(FileSystem sourceFs, List<ReplChangeManager.FileInfo> srcFileList,
                            FileSystem destinationFs, Path destination,
                            boolean useRegularCopy) throws IOException, LoginException {
     int repeat = 0;
-    List<Path> pathList = Lists.transform(fileList, ReplChangeManager.FileInfo::getEffectivePath);
+    boolean isCopyError = false;
+    List<Path> pathList = Lists.transform(srcFileList, ReplChangeManager.FileInfo::getEffectivePath);
```
Comment (maheshk114) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java]: if we do transform here ..is it possible that some of the entries will be null ?
may be transform is not required here ?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java [hunk 3]
```diff
@@ -82,71 +82,126 @@ public void copyAndVerify(FileSystem destinationFs, Path destRoot,
...
-              // If checksum does not match, likely the file is changed/removed, copy again from cm
-              srcFile.setIsUseSourcePath(false);
-              retryFileInfoList.add(srcFile);
-            }
-          }
-        }
-        if (!retryFileInfoList.isEmpty()) {
-          doCopyRetry(sourceFs, retryFileInfoList, destinationFs, destination, useRegularCopy);
-        }
+		    // Copy files with retry logic on failure
```
Comment (maheshk114) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java]: retry logic on failure and source file modification

File: hcatalog/server-extensions/src/main/java/org/apache/hive/hcatalog/listener/DbNotificationListener.java [hunk 4]
```diff
@@ -426,7 +426,13 @@ public boolean hasNext() {
 
     @Override
     public String next() {
-      String result = encodeFileUri(files.get(i), chksums != null? chksums.get(i) : null);
+      String result;
+      try {
+        result = ReplChangeManager.encodeFileUri(files.get(i), chksums != null ? chksums.get(i) : null, null);
+      } catch (IOException e) {
+        // If thrown IOException, it means, CM root is not set. So, just return original path.
+        result = files.get(i);
```
Comment (maheshk114) [inline hcatalog/server-extensions/src/main/java/org/apache/hive/hcatalog/listener/DbNotificationListener.java]: i think .. ReplChangeManager should handle (null check is already there) ..cm_root not set scenario ..here in this layer we should throw exception for any failure

File: standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/utils/FileUtils.java [hunk 5]
```diff
@@ -454,4 +455,59 @@ public static boolean isDirEmpty(FileSystem fs, Path path) throws IOException {
...
+    String authority = pathUri.getAuthority();
+
+    // validate/fill-in scheme and authority. this follows logic
+    // identical to FileSystem.get(URI, conf) - but doesn't actually
+    // obtain a file system handle
+
+    if (scheme == null) {
+      // no scheme - use default file system uri
+      scheme = fsUri.getScheme();
+      authority = fsUri.getAuthority();
```
[the remark this answers is not in the document]
Comment (sankarh) [reply standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/utils/FileUtils.java]: What is different in this logic?

File: standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/utils/StringUtils.java [hunk 6]
```diff
@@ -109,4 +109,34 @@ public static String stringifyException(Throwable e) {
...
+   * @param bytes Input bytes
+   * @param start start index, inclusively
+   * @param end end index, exclusively
+   * @return hex string representation of the byte array
+   */
+  public static String byteToHexString(byte[] bytes, int start, int end) {
+    if(bytes == null) {
+      throw new IllegalArgumentException("bytes == null");
+    } else {
+      StringBuilder s = new StringBuilder();
```
[the remark this answers is not in the document]
Comment (sankarh) [reply standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/utils/StringUtils.java]: Conflicting StringUtils classes if I import hadoop's class. So, just moved the implementation here as it is standalone metastore and even other methods does the same.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java [hunk 7]
```diff
@@ -82,71 +82,126 @@ public void copyAndVerify(FileSystem destinationFs, Path destRoot,
...
       }
-      repeat++;
+      srcPath = srcFile.getEffectivePath();
+      if (null == srcPath) {
+        // This case possible if CM path is not enabled.
+        LOG.error("File copy failed and likely source file is deleted or modified. "
+                + "Source File: " + srcFile.getSourcePath());
+        throw new IOException("File copy failed and likely source file is deleted or modified.");
+      }
+      pathList.add(srcFile.getEffectivePath());
```
[the remark this answers is not in the document]
Comment (sankarh) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java]: srcPath won't match with srcFile.getEffectivePath() as we changed the isUseSrcFile flag.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java (hunk 1 above)
Comment (sankarh) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java]: not possible. I'm checking for pathless not empty or copy error, then retry.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java (hunk 2 above)
Comment (sankarh) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java]: nope... here it won't return null. but just traverse through the list and get the effective path. So, cannot skip the transform.

File: hcatalog/server-extensions/src/main/java/org/apache/hive/hcatalog/listener/DbNotificationListener.java (hunk 4 above)
Comment (sankarh) [reply hcatalog/server-extensions/src/main/java/org/apache/hive/hcatalog/listener/DbNotificationListener.java]: This method is inherited and cannot change the definition. So, exception is caught and handle it. Yes, it is dead code for now.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java (hunk 3 above)
Comment (sankarh) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java]: Will fix this if any code change needed for this patch. :)

File: standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/utils/StringUtils.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sankarh) [reply standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/utils/StringUtils.java]: Done

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (sankarh) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java]: Done

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java (hunk 1 above)
Comment (sankarh) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java]: done

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java (hunk 3 above)
Comment (sankarh) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java]: done

File: hcatalog/server-extensions/src/main/java/org/apache/hive/hcatalog/listener/DbNotificationListener.java (hunk 4 above)
Comment (sankarh) [reply hcatalog/server-extensions/src/main/java/org/apache/hive/hcatalog/listener/DbNotificationListener.java]: done

**Pull request closed without merging**